### DISCUSSION TITLE
Dynamic Nodes and Choice Nodes

### DISCUSSION THREAD

**Discussion by evesdropper** [L3MON4D3/LuaSnip#712] [Q&A]
I apparently can't read and think because I implemented my previous snippet a little bit incorrectly :sweat_smile: - I realized I would also want to account for iterated integrals. Here's my previous snippet.

```lua
local int2 = function(args, snip)
    local vars = tonumber(snip.captures[1])
    local nodes = {}
    for j = 1, vars do 
        table.insert(nodes, t" \\dd ")
        table.insert(nodes, r(j, "var"..tostring(j), i(1)))
    end
    return sn(nil, nodes)
end

autosnippet({ trig='(%d)int', name='multi integrals', dscr='please work', regTrig=true, hidden=true},
    fmt([[ 
    \<><>nt_{<>}^{<>} <> <> <>
    ]],
    { c(1, {t(""), t("o")}), f(function(_, snip)
        inum = tonumber(snip.captures[1])
        res = string.rep("i", inum)
        return res
        end), i(2, "-\\infty"), i(3, "\\infty"), i(4), d(5, int2), i(0)},
    { delimiters='<>' }
    ), { condition=math, show_condition=math }), 
```

I've slightly modified it to introduce the following choice:
- Iterated Integrals: it'll generate `n` copies of `\int_{<>}^{<>}`
- Multiple Integrals: it'll generate a snippet node with `<>nt_{<>}`, where `<>` repeats `i` `n` times

This means a new function and a new choice node snippet:

```lua
-- generate number of \int_{<>}^{<>}
local int1 = function(args, snip)
    local vars = tonumber(snip.captures[1])
    local nodes = {}
    for j = 1, vars do 
        table.insert(nodes, t"\\int_{")
        table.insert(nodes, r(j, "lb"..tostring(j), i(1)))
        table.insert(nodes, t"}^{")
        table.insert(nodes, r(j, "ub"..tostring(j), i(1)))
        table.insert(nodes, t"} ")
    end
    return sn(nil, nodes)
end 

-- generate number of \dds
local int2 = function(args, snip)
    local vars = tonumber(snip.captures[1])
    local nodes = {}
    for j = 1, vars do 
        table.insert(nodes, t" \\dd ")
        table.insert(nodes, r(j, "var"..tostring(j), i(1)))
    end
    return sn(nil, nodes)
end

autosnippet({ trig='(%d)test', name='multi integrals', dscr='please work', regTrig=true, hidden=true},
    fmt([[ 
    <> <> <> <>
    ]],
    { c(1, { 
    fmta([[
    \<><>nt_{<>}
    ]], {c(1, {t(""), t("o")}), f(function(_, parent, snip)
        inum = tonumber(parent.parent.captures[1])  -- this guy's lineage looking like a research lab's 
        res = string.rep("i", inum)
        return res
        end), i(2)}), d(nil, int1)}) 
    , i(2), d(3, int2), i(0)},
    { delimiters='<>' }
    ), { condition=math, show_condition=math }),

```

The `fmta` choice works fine (except for the minor irritation that you have to tab out before you can switch to the next choice, but I have been having difficulty with the dynamic node. I look a look [here](https://github.com/L3MON4D3/LuaSnip/discussions/358) but it doesn't seem like the issue is the same. 

![image](https://user-images.githubusercontent.com/82856360/210155378-3235df27-f536-4be3-9da5-78abae397bcb.png)

Any idea on how to fix it (and with indexing dynamic nodes in choice nodes in general)? It doesn't seem like there is information on that.

A lot of consideration for a snippet that I likely won't even use for a while :rofl: Guess that's what time off from school is for.

**Comment by L3MON4D3** [comment]
> A lot of consideration for a snippet that I likely won't even use for a while 🤣 Guess that's what time off from school is for.
 
First of all, obligatory [xkcd](https://xkcd.com/1205) :P
(But I can definitely relate xD)

I'm not at my PC at the moment, could you share a formatted version of the snippet at the bottom? I can follow which node belongs where xD
Apart from this, int1 generates a snippet with some duplicated jump-indices, that might also lead to issues, but doesn't seem like it's the source of this one

Very nice ideas though, if you finish the snippets, would you add them to the wiki?? :)

**Reply by evesdropper** [reply]
> I'm not at my PC at the moment, could you share a formatted version of the snippet at the bottom? I can follow which node belongs where xD

Not exactly sure how I can easily do so :sob: but here's an attempt by hand:

`<1> <2> <3> <4>`

- 1: Choice Node
    - Option 1: Multiple Integrals - Fmt node, insert `\<fn>nt_{<ins>}`, where the function node inserts `i` number times.
    - Option 2: Iterated Integrals - Dynamic Node, insert number of `\int_{<ins>}^{<ins>} ` depending on number entered
- 2: Insert Node, add  expression to be evaluated
- 3: Dynamic Node, insert number of `\dd <insert>` text-insert pairs depending on number entered
- 4: Insert Node, extra insert at the end for other expressions 

Tried using the preview and I think that might be a reason why stuff is wonky - seems like something is cancelling out the `$` at 1? Maybe it would be alleviated if I used a snippet node but made things a bit messier.
![image](https://user-images.githubusercontent.com/82856360/210183094-8ac01294-4063-4425-ac48-dca0c0b6a790.png)


> Apart from this, int1 generates a snippet with some duplicated jump-indices, that might also lead to issues, but doesn't seem like it's the source of this one

I guess we'll cross that bridge when we get there. I assume that the restore nodes should be `i(1)` and `i(2)` then? Otherwise not too sure about repeats.

> Very nice ideas though, if you finish the snippets, would you add them to the wiki?? :)

Yup - I'm planning on adding this and another snippet once I get another feature of that working - stay tuned :eyes:

**Reply by L3MON4D3** [reply]
```lua
-- generate number of \int_{<>}^{<>}
local int1 = function(args, snip)
    local vars = tonumber(snip.captures[1])
    local nodes = {}
    for j = 1, j do
        table.insert(nodes, t"\\int_{")
        table.insert(nodes, r(2*j-1, "lb"..tostring(j), i(1)))
        table.insert(nodes, t"}^{")
        table.insert(nodes, r(2*j, "ub"..tostring(j), i(1)))
        table.insert(nodes, t"} ")
    end
    return sn(nil, nodes)
end

-- generate number of \dds
local int2 = function(_, snip)
    local vars = tonumber(snip.captures[1])
    local nodes = {}
    for j = 1, vars do
        table.insert(nodes, t" \\dd ")
        table.insert(nodes, r(j, "var"..tostring(j), i(1)))
    end
    return sn(nil, nodes)
end

ls.snip_expand(s({ trig='(%d)test', name='multi integrals', dscr='please work', regTrig=true, hidden=true},
    fmt([[ 
    <> <> <> <>
    ]],
    {
    	c(1, {
			fmta([[
				\<><>nt_{<>}
				]], {
					c(1, {t(""), t("o")}),
					f(function(_, parent)
						local inum = tonumber(parent.snippet.captures[1])  -- this guy's lineage looking like a research lab's 
						local res = string.rep("i", inum)
						return res
					end),
					i(2)}),
			d(nil, int1)
		}),
		i(2),
		d(3, int2),
		i(0)
	},
	{ delimiters='<>' }),
	{ condition=math, show_condition=math }),
	{expand_params={captures={"3"}}}  )
```
Okay,  I think it was just due to the indices of the restoreNodes xD

**Reply by evesdropper** [reply]
Yup haha - I've got to make a note to keep track of those :rofl: 

Welp that means it's time to post this snip onto the Wiki now that it works

**Reply by L3MON4D3** [reply]
Oh niiice, first usage I've seen of condition-objects, very sleek :D
